### PR TITLE
[engine] Transmit completed transaction via SResultFinalValue.

### PR DESCRIPTION
**Pull request by nickchapman-da**
Refactor to transmit completed transaction along with final `SValue` inside `SResultFinalValue`.

- extend `SResultFinalValue` to transmit completed transaction
- remove `OnLedger.finish`
- distinguish `SpeedyComplete` from `SpeedyHungry`
- remove need for `PartialTransaction.IncompletedTransaction`
- `PartialTransaction.finish` is now `private[speedy]`

During this refactor, the following issue/bug was discovered: https://github.com/digital-asset/daml/issues/14431

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 1]
```diff
@@ -137,7 +137,7 @@ private[lf] object Speedy {
       else {
         SVisibleToStakeholders.fromSubmitters(committers, readAs)
       }
-    private[lf] def finish: PartialTransaction.Result = ptx.finish
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:140]: payoff from this change

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala [hunk 2]
```diff
@@ -198,15 +198,14 @@ private[lf] object PartialTransaction {
...
-  sealed abstract class Result extends Product with Serializable
-  final case class CompleteTransaction(
+  private[lf] final case class Result(
       tx: SubmittedTransaction,
       locationInfo: Map[NodeId, Location],
       seeds: NodeSeeds,
       globalKeyMapping: Map[GlobalKey, KeyMapping],
       disclosedContracts: ImmArray[DisclosedContract],
-  ) extends Result
-  final case class IncompleteTransaction(ptx: PartialTransaction) extends Result
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala:209]: payoff from this change

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala [hunk 3]
```diff
@@ -312,25 +311,25 @@ private[speedy] case class PartialTransaction(
     * - an error in case the transaction cannot be serialized using
     *   the `outputTransactionVersions`.
     */
-  def finish: PartialTransaction.Result =
+  private[speedy] def finish: PartialTransaction.Result =
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala:314]: payoff from this change

**Review by remyhaemmerle-da — APPROVED** [review]
Nice cleanup.
Thanks.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 4]
```diff
@@ -1638,14 +1641,21 @@ private[lf] object Speedy {
   }
 
   /** Internal exception thrown when a continuation result needs to be returned.
-    *    Or machine execution has reached a final value.
     */
   private[speedy] final case class SpeedyHungry(result: SResult)
       extends RuntimeException
       with NoStackTrace {
     override def toString = s"SpeedyHungry($result)"
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala]: removed

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala [hunk 5]
```diff
@@ -198,15 +198,14 @@ private[lf] object PartialTransaction {
...
+  private[lf] final case class Result(
       tx: SubmittedTransaction,
       locationInfo: Map[NodeId, Location],
       seeds: NodeSeeds,
       globalKeyMapping: Map[GlobalKey, KeyMapping],
       disclosedContracts: ImmArray[DisclosedContract],
-  ) extends Result
-  final case class IncompleteTransaction(ptx: PartialTransaction) extends Result
+  ) extends Product
+      with Serializable
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala]: removed

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala [hunk 6]
```diff
@@ -20,8 +20,11 @@ sealed abstract class SResult extends Product with Serializable
 object SResult {
   final case class SResultError(err: SError) extends SResult
 
-  /** The speedy machine has completed evaluation to reach a final value. */
-  final case class SResultFinalValue(v: SValue) extends SResult
+  /** The speedy machine has completed evaluation to reach a final value.
+    * And, if the evaluation was on-ledger, a completed transaction.
+    */
+  final case class SResultFinalValue(v: SValue, tx: Option[PartialTransaction.Result])
+      extends SResult
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala]: renamed as suggested

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 7]
```diff
@@ -519,13 +519,16 @@ private[lf] object Speedy {
...
+        case SpeedyComplete(value: SValue) =>
+          if (enableInstrumentation) track.print()
+          ledgerMode match {
+            case OffLedger => SResultFinalValue(value, None)
+            case onLedger: OnLedger =>
+              val ctx = onLedger.ptx.finish
+              SResultFinalValue(value, Some(ctx))
           }
+
+        case SpeedyHungry(res: SResult) =>
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala]: swapped

**Pull request merged**